### PR TITLE
Boost thread example

### PR DESCRIPTION
**Pull request by tatraian**
Simple test application for boost thread

### CODE REVIEW

File: examples/Boost-thread/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# Emulate HunterGate:
+# * https://github.com/hunter-packages/gate
+include("../common.cmake")
+
+project(download-boost)
+
+# download boost
+hunter_add_package(Boost COMPONENTS system thread)
+
+find_package( Threads )
```
Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: There is no explicit use of this package. If it's needed for Boost then it should be done in `BoostConfig.cmake`.

File: examples/Boost-thread/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+# now boost can be used
+find_package(Boost CONFIG REQUIRED system thread)
+add_executable(foo foo.cpp)
+target_link_libraries(foo Boost::system Boost::thread ${CMAKE_THREAD_LIBS_INIT})
```
Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: Same logic as above for `CMAKE_THREAD_LIBS_INIT`

**Comment by tatraian** [comment]
I don't know why that one test failed.

**Comment by ruslo** [comment]
> I don't know why that one test failed.

Not a problem, just a Travis glitch. I've restarted job.

Please take a look at my comments above.

**Comment by tatraian** [comment]
I modified `BoostConfig.cmake` I hope this is OK!

File: cmake/templates/BoostConfig.cmake.in [hunk 3]
```diff
@@ -102,3 +102,12 @@ if(TARGET Boost::iostreams)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostZLIBDependency.cmake" OPTIONAL)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostBZip2Dependency.cmake" OPTIONAL)
 endif()
+
+if(TARGET Boost::thread)
+  find_package(Threads)
+  set_target_properties(
+    Boost::thread
+    PROPERTIES
+    INTERFACE_LINK_LIBRARIES "${CMAKE_THREAD_LIBS_INIT}"
```
Comment (ruslo) [inline cmake/templates/BoostConfig.cmake.in]: We already have `set_target_properties(... PROPERTIES INTERFACE_LINK_LIBRARIES Boost::boost")` so I think we should append property (read, modify, set).

Comment (ruslo) [inline cmake/templates/BoostConfig.cmake.in]: Also probably we should use [Threads::Threads](https://cmake.org/cmake/help/latest/module/FindThreads.html).

File: cmake/templates/BoostConfig.cmake.in [hunk 4]
```diff
@@ -102,3 +102,12 @@ if(TARGET Boost::iostreams)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostZLIBDependency.cmake" OPTIONAL)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostBZip2Dependency.cmake" OPTIONAL)
 endif()
+
+if(TARGET Boost::thread)
+  find_package(Threads)
```
Comment (ruslo) [inline cmake/templates/BoostConfig.cmake.in]: If we are using it without `REQUIRED` then there must be check `if(*_FOUND)` somewhere? It will be easier to add `REQUIRED` though.

**Comment by tatraian** [comment]
I hope I fixed both of them.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.912
* https://travis-ci.org/ingenue/hunter/builds/157663060

File: examples/Boost-thread/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+cmake_minimum_required(VERSION 3.0)
+
+# Emulate HunterGate:
+# * https://github.com/hunter-packages/gate
+include("../common.cmake")
+
+project(download-boost)
+
+# download boost
+hunter_add_package(Boost COMPONENTS system thread)
```
Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: We are using `Boost::chrono` in fact, see foo.cpp. I guess that's the reason why test fail on Windows: https://ci.appveyor.com/project/ingenue/hunter/build/1.0.912/job/g631xocheg8xhsd9

Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: Actually I vote for not mixing unrelated stuff. E.g. we can just print thread-id to std::cout, no need to use chrono.

File: cmake/templates/BoostConfig.cmake.in [hunk 6]
```diff
@@ -102,3 +102,11 @@ if(TARGET Boost::iostreams)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostZLIBDependency.cmake" OPTIONAL)
   include("${CMAKE_CURRENT_LIST_DIR}/BoostBZip2Dependency.cmake" OPTIONAL)
 endif()
+
+if(TARGET Boost::thread)
+  find_package(Threads REQUIRED)
+  set_property(
+    TARGET "Boost::thread"
+    APPEND PROPERTY INTERFACE_LINK_LIBRARIES Threads::Threads
```
Comment (ruslo) [inline cmake/templates/BoostConfig.cmake.in]: Error on Linux reproduced now:
* https://s3.amazonaws.com/archive.travis-ci.org/jobs/157663063/log.txt
```
thread.cpp:(.text+0x2c5): undefined reference to `clock_gettime'
```

Can be fixed by adding `-lrt`: http://stackoverflow.com/a/2418175/2288008
In CMake terms this will be:
```cmake
include(CheckLibraryExists) # check_library_exists
check_library_exists(rt clock_gettime "" _boost_check_rt_library)
if(_boost_check_rt_library)
  # Add library to 'INTERFACE_LINK_LIBRARIES' of Boost::thread target
endif()
```

File: examples/Boost-thread/CMakeLists.txt (hunk 5 above)
Comment (tatraian) [inline examples/Boost-thread/CMakeLists.txt]: Ok. But since static linking we won't get the link error if this_thread::id is not depends on librt. I don't know this.

Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: > we won't get the link error if this_thread::id is not depends on librt

Add Boost::chrono example :)

Comment (tatraian) [inline examples/Boost-thread/CMakeLists.txt]: Of course... This is the situation when I don't use my brain... Can I add that to this pull request or you would like an other?

Comment (tatraian) [inline examples/Boost-thread/CMakeLists.txt]: This fixes the Threads::Threads dependency of boost::thread

Comment (ruslo) [inline examples/Boost-thread/CMakeLists.txt]: > Can I add that to this pull request or you would like an other?

No, let's finish this first. Who knows what troubles we will have with chrono.

**Comment by tatraian** [comment]
I removed chrono, but I inspected `boost::thread` code and I found that it contains `clock_gettime` call:
https://github.com/boostorg/thread/blob/develop/include/boost/thread/pthread/timespec.hpp#L84

I try to figure out a test scenario for that call...

**Comment by tatraian** [comment]
Anyway, `chrono` also uses `clock_gettime` so the error will be fixed there as well.

If you think we can close this pr here and can start a new for `rt` library.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.915
* https://travis-ci.org/ingenue/hunter/builds/157828738

**Comment by ruslo** [comment]
Uploading:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.917
* https://travis-ci.org/ingenue/hunter/builds/157857998

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.15.25
